File: 3568-minimum-moves-to-clean-the-classroom/3568-minimum-moves-to-clean-the-classroom.cpp
class Solution {
public:
    int minMoves(vector<string>& classroom, int energy) {
        int m = classroom.size();
        int n = classroom[0].size();

        int startR = 0, startC = 0;

        // Give every litter cell a bit number
        vector<vector<int>> litterId(m, vector<int>(n, -1));
        int litterCount = 0;

        for (int i = 0; i < m; i++) {
            for (int j = 0; j < n; j++) {

                if (classroom[i][j] == 'S') {
                    startR = i;
                    startC = j;
                }

                if (classroom[i][j] == 'L') {
                    litterId[i][j] = litterCount++;
                }
            }
        }

        int finalMask = (1 << litterCount) - 1;

        // State: row, col, remaining energy, mask, distance
        queue<tuple<int, int, int, int, int>> q;

        q.push({startR, startC, energy, 0, 0});

        // visited[row][col][energy][mask]
        vector<vector<vector<vector<bool>>>> visited(
            m,
            vector<vector<vector<bool>>>(
                n,
                vector<vector<bool>>(
                    energy + 1,
                    vector<bool>(1 << litterCount, false)
                )
            )
        );

        visited[startR][startC][energy][0] = true;

        int dr[] = {-1, 1, 0, 0};
        int dc[] = {0, 0, -1, 1};

        while (!q.empty()) {

            auto [r, c, currEnergy, mask, dist] = q.front();
            q.pop();

            // All litter collected
            if (mask == finalMask) {
                return dist;
            }

            for (int d = 0; d < 4; d++) {

                int nr = r + dr[d];
                int nc = c + dc[d];

                // Outside grid
                if (nr < 0 || nr >= m ||
                    nc < 0 || nc >= n) {
                    continue;
                }

                // Obstacle
                if (classroom[nr][nc] == 'X') {
                    continue;
                }

                // Cannot move if no energy
                if (currEnergy == 0) {
                    continue;
                }

                int newEnergy = currEnergy - 1;
                int newMask = mask;

                // Collect litter
                if (classroom[nr][nc] == 'L') {
                    int id = litterId[nr][nc];
                    newMask |= (1 << id);
                }

                // Reset energy
                if (classroom[nr][nc] == 'R') {
                    newEnergy = energy;
                }

                if (!visited[nr][nc][newEnergy][newMask]) {

                    visited[nr][nc][newEnergy][newMask] = true;

                    q.push({
                        nr,
                        nc,
                        newEnergy,
                        newMask,
                        dist + 1
                    });
                }
            }
        }

        return -1;
    }
};

File: 3568-minimum-moves-to-clean-the-classroom/3568-minimum-moves-to-clean-the-classroom.py
from collections import deque

class Solution:
    def minMoves(self, classroom: list[str], energy: int) -> int:

        m = len(classroom)
        n = len(classroom[0])

        startR = 0
        startC = 0

        # Give every litter an ID
        litterId = [[-1] * n for _ in range(m)]
        litterCount = 0

        for i in range(m):
            for j in range(n):

                if classroom[i][j] == 'S':
                    startR = i
                    startC = j

                if classroom[i][j] == 'L':
                    litterId[i][j] = litterCount
                    litterCount += 1

        finalMask = (1 << litterCount) - 1

        # State:
        # (row, col, energy, mask, distance)
        q = deque()

        q.append((startR, startC, energy, 0, 0))

        # visited[row][col][energy][mask]
        visited = set()

        visited.add((startR, startC, energy, 0))

        directions = [
            (-1, 0),
            (1, 0),
            (0, -1),
            (0, 1)
        ]

        while q:

            r, c, currEnergy, mask, dist = q.popleft()

            # All litter collected
            if mask == finalMask:
                return dist

            for dr, dc in directions:

                nr = r + dr
                nc = c + dc

                # Outside grid
                if nr < 0 or nr >= m or nc < 0 or nc >= n:
                    continue

                # Obstacle
                if classroom[nr][nc] == 'X':
                    continue

                # No energy to make a move
                if currEnergy == 0:
                    continue

                newEnergy = currEnergy - 1
                newMask = mask

                # Collect litter
                if classroom[nr][nc] == 'L':
                    litter = litterId[nr][nc]
                    newMask |= (1 << litter)

                # Reset energy
                if classroom[nr][nc] == 'R':
                    newEnergy = energy

                state = (nr, nc, newEnergy, newMask)

                if state not in visited:

                    visited.add(state)

                    q.append((
                        nr,
                        nc,
                        newEnergy,
                        newMask,
                        dist + 1
                    ))

        return -1